### PR TITLE
Add a first pass at a course overview

### PR DESCRIPTION
**Pull request by nickstenning**
Adds a first-pass course overview in terms of four workshops, with clear learning goals for each one.

Things to note:

- from my perspective this is more an exercise in finding the right structure and tone than in finding exactly the right content
- I'm open to feedback on any aspect of this document

Questions for reviewers:

- Is the tone right? Too light? Too serious?
- Is the scope for each 2h workshop too much? (Clearly this is going to be hard to answer without more detail on each workshop, but the important question is: will it be possible to design _some_ course materials that address the listed learning areas that fits into 2h?)

Thanks in advance!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,109 @@
...
 
-1. [Week 1](./week1-webclient.md)
+## Introduction
+
+The course is going to be structured as a series of workshops, during which
+you will learn about various technologies and practices used in modern
+software development.
+
+In each workshop, you and your peers will be the development team working on
+part of a web application for a pet adoption center. You will learn about
```
Comment (notheotherben) [inline README.md]: nit: Should we use the term "animal shelter" here rather than "pet adoption center"? (which sounds a lot more 1984) :smile:

File: README.md [hunk 2]
```diff
@@ -1,5 +1,109 @@
...
+front-end web development, APIs, testing and continuous integration, as well
+as how to deploy web applications to the cloud, and how to collaborate in a
+modern software team.
+
+During the workshop, you will have access to a group of mentors, who will be
```
Comment (notheotherben) [inline README.md]: nit: I'd suggest removing the second comma here, generally when commas are used in pairs they indicate that the text between them may be skipped and the sentence should continue to read naturally.

File: README.md [hunk 3]
```diff
@@ -1,5 +1,109 @@
...
+
+## Learning goals
+
+Our goal is to give you a chance to learn about modern software development
+practices, and especially to learn about how web applications are designed,
+built, tested, and deployed.
+
+Over the duration of the course, you will build all the important parts of a
+web application that allows users to find and adopt adorable furry
+companions. Each workshop will focus on a different part of the application,
```
Comment (notheotherben) [inline README.md:31]: I :heart: the tone here, this is meant to be as fun as it is educational (in my view, we're trying to build a community after all) and "adorably furry companions" definitely sets the tone well.

File: README.md [hunk 4]
```diff
@@ -1,5 +1,109 @@
...
+and will draw on a different set of skills.
+
+Below, you'll find out what to expect in each workshop.
+
+## Course outline
+
+### Workshop 1: modern front-end web development
```
Comment (notheotherben) [inline README.md]: nit: Might be worth using title case here to match the other titles 🤔

File: README.md [hunk 5]
```diff
@@ -1,5 +1,109 @@
...
+- How to create a SPA using `create-react-app`.
+- How to scaffold a user interface using `react-bootstrap`.
+- What is an API (application programming interface)?
+- How to talk to an API from our React SPA.
+
+We'll also cover:
+
+- How to use the Git version control system to submit a pull request (PR).
+- How to use pull requests as part of a code review process.
+
```
Comment (notheotherben) [inline README.md:48]: Do you think it is worth having an entry here for "whiteboarding/wireframing a UI design based on user requirements"?

File: README.md [hunk 6]
```diff
@@ -1,5 +1,109 @@
...
+_If you're ready, go to the [course notes for Workshop 1](./week1-webclient.md)._
+
+### Workshop 2: APIs and back-end web development
+
+In the first workshop, you built a single-page application that talked to an
+API service that returned information about pets up for adoption. In this
+workshop, you're going to build the pet adoption API yourselves, and host it
+in the cloud. If things go according to plan, you'll learn:
```
Comment (notheotherben) [inline README.md]: I'd leave out the "if things go according to plan" - I think it comes across as a bit dismissive and unprepared. Hopefully by this stage, participants will already have experience working on this project for a decent duration and won't be afraid of spending a few hours here and there outside of the workshop learning more about it.

File: README.md [hunk 7]
```diff
@@ -1,5 +1,109 @@
...
+
+- How is an API application different from other web applications?
+- How to pick between different options for building and hosting web applications.
+- How to build an API for listing pets available for adoption.
```
Comment (notheotherben) [inline README.md]: When you say "build an API" do you specifically mean "implementation" or the "design" of the API? If the former, maybe it's worth adding an entry for the latter - since API design is one of those subtle things which is really good to go over.

File: README.md [hunk 8]
```diff
@@ -1,5 +1,109 @@
...
+- How to host your API application in Azure.
+
+_The notes for Workshop 2 aren't quite ready yet. We'll add them here when
+they are._
+
+### Workshop 3: Testing, Continuous Integration, Continuous Deployment
+
+So far, you've built both a front-end and a back-end for the pet adoption
+center's website. In this workshop, we're going to look at how you can test
+the changes you make to your software, and how you can integrate those tests
```
Comment (notheotherben) [inline README.md]: I'd hope that our participants have been writing tests all along and that testing isn't something that they tack-on afterwards. As such, I'd suggest phrasing this as specifically "how do we automate the running of tests by setting up an automated test runner on Azure, deploy automatically and how to map your release process to your Git process".

File: README.md [hunk 9]
```diff
@@ -1,5 +1,109 @@
...
+
+_The notes for Workshop 3 aren't quite ready yet. We'll add them here when
+they are._
+
+### Workshop 4: Logs, Metrics, and Observability
+
+You have now built, tested, and deployed a web application to the cloud! Now
+you want to know whether it's doing the right thing. In this workshop, we're
+going to look at how to use logs and metrics to understand what your
+application is really doing in production. By adding instrumentation to your own application, you'll learn:
```
Comment (notheotherben) [inline README.md]: This all sounds great, but the SRE in me wants to sell the idea of "We'd like to make sure that your **customers** are able to happily use your service". What do you think?

File: README.md [hunk 10]
```diff
@@ -1,5 +1,109 @@
...
+
+_The notes for Workshop 4 aren't quite ready yet. We'll add them here when
+they are._
+
+## Contributions
+
+All of our course notes are open-source, and we invite corrections and
+contributions and corrections from anyone. Find a typo? [Send us a pull
+request](pull/new). Want to contribute a new module? [Open an issue to
+discuss your idea][issues/new].
```
Comment (notheotherben) [inline README.md]: Should this be `(issues/new)` rather than `[issues/new]`?

**Review by notheotherben — COMMENTED** [review]
That looks great Nick, really great first pass which I think sets things on the right course (😉). One thing I think would be good to change would be to move testing further up (i.e. include testing in each section) and then leave section 3 for how we automate and leverage that to simplify our lives rather than asking participants to add tests to code they've already written (which I personally find challenging even in the best of cases).

As for the tone, I really like it - it is succinct, light hearted but serious about the technical aspects - essentially what I hope the course can be. 

In terms of how much content there is, you could basically make a career out of any of those sections if you wanted to - I suspect we can make them as dense or sparse as we want. I really like the idea of us doing almost "lightning talks" on each of the high level topics within a section (say, 5 minutes) to give a brief intro to them and then get hands on experience. I also suspect that much of the depth will happen over the course of the next `$time_span` after the workshop if we can do a good job of fostering a community which enables that - in which case our engagement there will likely be more important.

File: README.md [hunk 11]
```diff
@@ -1,5 +1,109 @@
...
+- What is a single-page application (SPA)?
+- How to create a SPA using `create-react-app`.
+- How to scaffold a user interface using `react-bootstrap`.
+- What is an API (application programming interface)?
+- How to talk to an API from our React SPA.
+
+We'll also cover:
+
+- How to use the Git version control system to submit a pull request (PR).
+- How to use pull requests as part of a code review process.
```
Comment (stirlaf) [inline README.md:47]: FYI, The process agenda items are possibly more important than the spa bits, because technology and libraries will vary by team and job, but these processes are common in all large scale companies.

**Review by stirlaf — APPROVED** [review]
Looks good.

File: README.md [hunk 12]
```diff
@@ -1,5 +1,109 @@
...
+
+- What is a single-page application (SPA)?
+- How to create a SPA using `create-react-app`.
+- How to scaffold a user interface using `react-bootstrap`.
+- What is an API (application programming interface)?
+- How to talk to an API from our React SPA.
+
+We'll also cover:
+
+- How to use the Git version control system to submit a pull request (PR).
```
Comment (lupino3) [inline README.md]: Nit, a PR is a GitHub concept, not a Git concept; moreover, folks are expected to have basic Git proficiency, so maybe `s/Git version control system/GitHub/` ?

File: README.md (hunk 5 above)
Comment (lupino3) [reply README.md:48]: Will it fit in the 2h?

Comment (notheotherben) [reply README.md:48]: I was thinking of running a quick 15 minute hands-on session - the initial UI doesn't need to be complicated at all (as Weslie mentioned before) and giving the class the opportunity to quickly mock that up before they start building it will probably foster more investment in the project than if this is given to them (and help avoid them getting stuck by not having a plan first).

File: README.md (hunk 1 above)
Comment (nickstenning) [reply README.md]: 😂👍

File: README.md (hunk 2 above)
Comment (nickstenning) [reply README.md]: 👍

File: README.md (hunk 4 above)
Comment (nickstenning) [reply README.md]: Yup. I'll make this consistent.

File: README.md (hunk 12 above)
Comment (nickstenning) [reply README.md]: Fair. Will tweak this.

File: README.md (hunk 11 above)
Comment (nickstenning) [reply README.md:47]: That's a good point. I'll shuffle these around a little and change the emphasis.

File: README.md (hunk 5 above)
Comment (nickstenning) [reply README.md:48]: Yeah, that's my concern. I think we can talk about it, but I'm not sure we're going to have time to do much more than a cursory exercise here.

File: README.md (hunk 7 above)
Comment (nickstenning) [reply README.md]: Will change this to "design and build".

File: README.md (hunk 8 above)
Comment (nickstenning) [reply README.md]: Yup, that's fair. For now I'm not going to make explicit reference to testing in earlier sessions, but we can build it into the course materials, and I'll refocus this one to be about automation, CI, CD, etc.

File: README.md (hunk 9 above)
Comment (nickstenning) [reply README.md]: Yup. I think for the pet shelter it may make sense to talk about "users" rather than customers, but I'll definitely reword this to frame that as the overall goal.

File: README.md (hunk 10 above)
Comment (nickstenning) [reply README.md]: Ah, nice catch. Thanks.

File: README.md (hunk 6 above)
Comment (nickstenning) [reply README.md]: :+1:

File: README.md (hunk 5 above)
Comment (nickstenning) [reply README.md:48]: Didn't see your latest before posting, Ben. Let's definitely see if we can fit something like this in, and we can later make a call about whether we're going to be too pressed for time.

**Comment by nickstenning** [comment]
Thanks everyone who reviewed this. I've pushed an updated version which I think should address most of the feedback. If anyone has time to give this one more once-over before merging, that'd be great.

File: README.md [hunk 13]
```diff
@@ -1,5 +1,104 @@
...
+
+### Workshop 1: Modern front-end web development
+
+In this workshop, we're going to build a frontend of a website for the pet
+adoption center. In doing this, you'll learn:
+
+- What is a single-page application (SPA)?
+- What is an API (application programming interface)?
+- How to talk to an API from a SPA built with React.
+- How to use the GitHub to submit a pull request (PR).
```
Comment (notheotherben) [inline README.md]: `s/the GitHub/GitHub/` :wink:

**Review by notheotherben — COMMENTED** [review]
Looks good to me

File: README.md (hunk 13 above)
Comment (nickstenning) [reply README.md]: lol. the GitHub. Love it.
